### ISSUE TITLE
Portable mode questions

### ISSUE THREAD

**Issue by sagikazarmark** [drakkan/sftpgo#199]
I'm trying to start sftpgo in portable mode (I really only need a single user, configuring it from the command line comes in handy). A couple questions came up though:

- Looks like I cannot change the configuration dir. Is that on purpose? (also, see #198)
- I'm a little bit confused if configuration would at all be loaded in portable mode. For instance, I'd like to change the path to host keys. Can I do that?
- ~The default values for `bind_port` values (eg. for ftpd). Config says `0 means disabled. Default: 0.`, but in the [code](https://github.com/drakkan/sftpgo/blob/b51d795e049e15df03495644d1b0f7f51206583d/service/service_portable.go#L71) there is `ftpPort >= 0` which means ftp will be started on a random port. Should I pass -1?~ Never mind, I found the answer in flags, but it's still a but confusing.
- Actually, is it a good idea to enable all services by default? I wonder if enabling them manually would be better. For example, I wouldn't want users to be able to use FTP EVER if it's up to me, so it's kind of a security issue by default for me.

Thanks!

**Comment by drakkan** [comment]
Hi,

portable mode by default enable sftp only:

```
./sftpgo portable
2020-10-20T09:29:55.091 INF Portable mode ready, SFTP port: 51359, user: "user", password: "kUaie4JX", public keys: [], directory: "/home/nicola/goprojects/sftpgo", permissions: map[/:[list download]], enabled ssh commands: [md5sum sha1sum cd pwd scp] file extensions filters: [] 
```

the configuration file is not [loaded](https://github.com/drakkan/sftpgo/blob/master/service/service.go#L61). 

Portable mode was designed to quickly share a folder not to be used as service. I can evaluate if I can extend it but I think that the best thing to do is to add some new flags to the `serve` subcommand to load a dump from a file, so you can load all the users you need at startup. I would like to expose via flags the [parameters](https://github.com/drakkan/sftpgo/blob/master/httpd/schema/openapi.yaml#L704)  for the loaddata REST API. This will also fix #161

**Comment by sagikazarmark** [comment]
I think at this point I'll just create a custom binary, it suits my use case, but might not work for everyone (eg. for licensing reasons as you pointed out in another issue).

I think running a (semi-)statless sftp server for uploading files (and possibly notifying other components in the system) is a perfectly valid use cases, especially in the cloud native era. Portable mode is the closest to supporting that use case at the moment.

> so you can load all the users you need at startup

I'm not sure how that would work, but unless it uses a file as a read-only data store (and not load users into a database for example), it's not really something I would run in a cloud native environment. I either want a service to be stateless, or I want to bootstrap the service. The service bootstrapping itself with some data brings up a whole lot of issues from running multiple instances to the eventual nature of dependencies (causing failures at startup).

**Comment by drakkan** [comment]
> I think at this point I'll just create a custom binary, it suits my use case, but might not work for everyone (eg. for licensing reasons as you pointed out in another issue).
> 
> I think running a (semi-)statless sftp server for uploading files (and possibly notifying other components in the system) is a perfectly valid use cases, especially in the cloud native era. Portable mode is the closest to supporting that use case at the moment.
> 
> > so you can load all the users you need at startup
> 
> I'm not sure how that would work, but unless it uses a file as a read-only data store (and not load users into a database for example),

You can use the memory provider using a file as read only data store, please take a look [here](https://github.com/drakkan/sftpgo/blob/master/docs/full-configuration.md) to the data_provider -> name.

> it's not really something I would run in a cloud native environment. I either want a service to be stateless, or I want to bootstrap the service. The service bootstrapping itself with some data brings up a whole lot of issues from running multiple instances to the eventual nature of dependencies (causing failures at startup).

**Comment by drakkan** [comment]
Please take a look [here](https://github.com/drakkan/sftpgo/commit/bb5207ad77a6b99e40de2367fef6de5ee19ad405), so you can:

1) load users on startup inside a memory provider and restore the same state each time the application start (this was already possibile as pointed above)
2) load users on startup using any supported provider. Obviously in this case if you start multiple instances simultaneously and you don't use an embedded provider (bolt or SQLite) you could get issues.

The portable mode basically create a user inside a memory provider, however users have a lot of options, the portable mode expose only a subset (to allow all the options we need almost 40 flags!). 

Loading one or more users from a dump allow to:

1) load more than one user
2) specify all the supported options

additionally:

1) using a memory provider you will have a stateless service you only need the dump file
2) if someone want to load some initial data inside a single instance this is now possibile. Initializing a fresh instance seems quite common for Docker containers. The new mode allows to load a dump and delete the file on success so the restore is executed only once. Alternately you can load users from dump only if they don't exist.

**Comment by sagikazarmark** [comment]
Thanks @drakkan, the loaddata+memory provider makes sense.

For now, I think I'll stick to the custom binary solution with the portable mode, mostly because of the embedded notification. The added complexity of an external binary (ensuring resiliency of an HTTP request or a command execution) just doesn't make sense in my use case. It's much easier to implement a backoff and logging in a custom action handler for now.

Thanks!

**Comment by sagikazarmark** [comment]
FYI I ended up switching back to external hooks and running the prebuilt SFTPGo Docker image, because I needed to support multiple users and it was easier just switch back than implementing a custom configuration layer for multiple users. The data loading is really useful for this purpose.

**Comment by drakkan** [comment]
Thank you for the update, for you info the portable mode should now [load the configuration file](https://github.com/drakkan/sftpgo/commit/f2acde789d66ed2c5e07771474fbce744bf1b9da#diff-1378cd41ba61c0a07d7dedb981d8ac36e627134475c94cb7954d1be55db97ff4R31), so hooks etc should work (honestly I didn't tested them in portable mode).

So we can now define a single, in memory user, using the portable mode or load a dump at startup (using any provider). Maybe I could bind portable flags with env vars but is a such tedious task

**Comment by sagikazarmark** [comment]
I think with the memory provider + loading the portable mode is probably less useful.
